### ISSUE TITLE
How to get left and right boundary of entire route(-map) object?

### ISSUE THREAD

**Issue by autonomobil** [fzi-forschungszentrum-informatik/Lanelet2#59]
I'm in the situation where I need the left and right boundary of a route object. Is this possible?

I generated the route object by calling:
 `Optional<lanelet::routing::Route> route =  routingGraph->getRoute(currentLanelet, targetLanelet, 0);`

Then I can generate a map-object from the route:
`LaneletMapConstPtr routeMap = route->laneletMap();`

Now I would like to have the left and right boundary of the entire route or the map object. What would be the best way to do this? 

I tried to take the shortestPath-object and look left and right for neighbour-lanelets, but then "merging"-relations are ignored. 

With `Optional<routing::RelationType> relation = routingGraph->routingRelation(lanelet1, lanelet2, false);` I can get those relations, but then it's getting complicated to decide which boundary from which lanelet to take.

**Comment by poggenhans** [comment]
Hm, your problem sounds simple at the first glance, but actually it isn't. Consider these two examples:

![image](https://user-images.githubusercontent.com/22795364/69061569-b37d1280-0a19-11ea-8a62-d0a3a158a34d.png)

If there was a generic notion of what the "left" and "right" lanelet is in a divering scenario, and crossing lanelets like in the first example can be ruled out, there might be a solution to this problem. In this case you can maybe use an algorithm that first computes the left and right bound ignoring merging lanelets and then close the gaps in the resulting shape that come from diverging lanelets by finding a path though the boundaries of the diverging lanelets that closes these gaps.

If you are only interested in the geometric shape of the left and right bound, you might think about using `boost::geometry` to calculate the polygon that is the union of all lanelets on the `Route` and then use that. Basically like this:

```
#include <lanelet2_core/primitives/Lanelet.h>
#include <lanelet2_core/geometry/Polygon.h>
#include <boost/geometry.hpp>

auto laneletUnion(const lanelet::Lanelets& llts) {
    using MultiPolygon = boost::geometry::model::multi_polygon<lanelet::BasicPolygonWithHoles2d>;
    MultiPolygon inputPoly;
    MultiPolygon outputPoly;
    for(auto& llt : llts) {
        boost::geometry::union_(inputPoly, llt.polygon2d().basicPolygon(), outputPoly);
        inputPoly = std::move(outputPoly);
    }
    return inputPoly.front();
}
```

**Comment by autonomobil** [comment]
Hey thanks for the fast response. 

I'm more interested in the shape of the boundary. With them I want to define the physically drivable space, rather than "pure" lanelet boundaries, which would trigger a replanning if we overtake another vehicle or obstacle and leave our shortestPath. 

I tried the boost::geometry part, but it won't compile:
```
/usr/include/boost/geometry/geometries/concepts/polygon_concept.hpp:64:85: error: invalid initialization of non-const reference of type ‘boost::geometry::concept::Polygon<lanelet::BasicPolygonWithHoles2d>::interior_mutable_type {aka std::vector<lanelet::BasicPolygon2d>&}’ from an rvalue of type ‘lanelet::BasicPolygons2d {aka std::vector<lanelet::BasicPolygon2d>}’
             interior_mutable_type i = traits::interior_rings<PolygonType>::get(*poly);
```
and
```
/usr/include/boost/geometry/core/interior_rings.hpp:83:27: error: invalid initialization of non-const reference of type ‘boost::geometry::interior_return_type<lanelet::BasicPolygonWithHoles2d>::type {aka std::vector<lanelet::BasicPolygon2d>&}’ from an rvalue of type ‘lanelet::BasicPolygons2d {aka std::vector<lanelet::BasicPolygon2d>}’
             >::get(polygon);
```
Unfortunately I'm not very good with boost::geometry ...

Even if I get this to work, how do I separate the polygon in left and right boundary again? Go along the centerline of the shortestPath and get the distance to the polygon left and right?

**Comment by poggenhans** [comment]
Just pull again. I fixed that short after my post :-D.

To separate the polygon, you could try to locate the first point in the boundary of the leftmost lanelet at the start of the route using `boost::geometry::equals` (and similarly for the end and the rightmost lanelet).

**Comment by autonomobil** [comment]
Thank you very much! 

I found a solution where I go along the shortestPath-object and check left and right for the outermost boundaries. The problem here is that merging/diverging/overlapping lanelets are not taken into account. But with the help of
 `using MultiPolygon = boost::geometry::model::multi_polygon<BasicPolygonWithHoles2d>;`  
and
`boost::geometry::union_()` 
and 
` boost::geometry::overlaps()` 
and your mentioned function I detect those overlapping lanelets, convert those lanelets to polygons and handle them separately.

The problem with converting the complete route to a polygon is with round-abouts. If your route goes back, then inner polygons get created, which complicates the topic so I chose the solution above.
